src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,48 +221,146 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// <summary>
+        /// Implementation of the reader logic when in state 'Start'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtStartImplementation();
 
-                case ODataBatchReaderState.Completed:
-                    // We should never get here when in Completed state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Completed));
+        /// <summary>
+        /// Implementation of the reader logic when in state 'Operation'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtOperationImplementation();
 
-                case ODataBatchReaderState.Exception:
-                    // We should never get here when in Exception state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Exception));
+        /// <summary>
+        /// Implementation of the reader logic when in state 'ChangesetStart'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtChangesetStartImplementation();
 
-                default:
-                    // Invalid enum value
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_UnknownValue));
+        /// <summary>
+        /// Implementation of the reader logic when in state 'ChangesetEnd'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtChangesetEndImplementation();
+
+        /// <summary>
+            /// Instantiate an <see cref="ODataBatchOperationRequestMessage"/> instance.
+        /// </summary>
+        /// <param name="streamCreatorFunc">The function for stream creation.</param>
+        /// <param name="method">The HTTP method used for this request message.</param>
+        /// <param name="requestUri">The request Url for this request message.</param>
+        /// <param name="headers">The headers for this request message.</param>
+        /// <param name="contentId">The contentId of this request message.</param>
+        /// <returns>The <see cref="ODataBatchOperationRequestMessage"/> instance.</returns>
+        protected ODataBatchOperationRequestMessage BuildOperationRequestMessage(
+            Func<Stream> streamCreatorFunc,
+            string method,
+            Uri requestUri,
+            ODataBatchOperationHeaders headers,
+            string contentId)
+        {
+            Uri uri = BuildOperationRequestUri(requestUri, this.inputContext.MessageReaderSettings.BaseUri);
+            this.contentIdToAddOnNextRead = contentId;

```

Comment:
Can we set this in CreateOperationRequestMessage, when we set the ReaderOperationState to MessageCreated? I'd rather do it there, when the derived class returns the final operation request message, than as a side effect of the derived class building the operation (which, technically, it could do at any time; we don't enforce state transitions, etc. (nor do we need to, if we move this line to CreatOperationRequestMessage).

Reply:
moved.

---
In reply to: [145847359](https://github.com/OData/odata.net/pull/960#discussion_r145847359) [](ancestors = 145847359)